python/pyarrow/tests/test_convert_builtin.py

```diff
@@ -488,6 +488,43 @@ def test_ndarray_nested_numpy_double(from_pandas, inner_seq):
                                 [[1., 2.], [1., 2., 3.], [np.nan], None])
 
 
+def test_nested_ndarray_in_object_array():
+    # ARROW-4350
+    arr = np.empty(2, dtype=object)
+    arr[:] = [np.array([1, 2]), np.array([2, 3])]
+
+    arr2 = np.empty(2, dtype=object)
+    arr2[0] = [3, 4]
+    arr2[1] = [5, 6]
+
+    expected_type = pa.list_(pa.list_(pa.int64()))
+    assert pa.infer_type([arr]) == expected_type
+
+    result = pa.array([arr, arr2])
+    expected = pa.array([[[1, 2], [2, 3]], [[3, 4], [5, 6]]],
+                        type=expected_type)
+
+    assert result.equals(expected)
+
```

Comment:
It might be worth adding explicitly the case of length-1 nested arrays, as this case changed compared to released version / master (it was doing something wrong before, see my [comment on the issue](https://issues.apache.org/jira/browse/ARROW-4350?focusedCommentId=16858438&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16858438))

This can look like:

```
# test case for len-1 arrays to ensure they are interpreted as sublists and not scalars
arr = np.empty(2, dtype=object)
arr[:] = [np.array([1]), np.array([2])]
result = pa.array([arr, arr])
assert result.to_pylist() == [[[1], [2]], [[1], [2]]]
```

I tested that this branch now gives the correct result, on master you would get:

```
In [74]: result.to_pylist()                                                                                                                                                                                        
Out[74]: [[[1, 2]], [[1, 2]]]
```

Reply:
done